tests/http_spec.rs

```diff
@@ -0,0 +1,216 @@
+    });
+    let server_context = ServerContext::new(blueprint, client);
+    Arc::new(server_context)
+  }
+}
+
+#[derive(Clone)]
+struct MockHttpClient {
+  upstream_mocks: Vec<(UpstreamRequest, UpstreamResponse)>,
+  expected_upstream_requests: Vec<UpstreamRequest>,
+}
+#[async_trait::async_trait]
+impl HttpClient for MockHttpClient {
+  async fn execute(&self, req: reqwest::Request) -> Result<Response, anyhow::Error> {
+    // Clone the mocks to allow iteration without borrowing issues.
+    let mocks = self.upstream_mocks.clone();
+
+    // Try to find a matching mock for the incoming request.
+    let mock = mocks
+      .iter()
+      .find(|(mock_req, _)| {
+        let method_match = req.method().as_str()
+          == serde_json::to_string(&mock_req.0.method.clone())
+            .expect("provided method is not valid")
+            .as_str()
+            .trim_matches('"');
+        let url_match = req.url().as_str() == mock_req.0.url.clone().as_str();
+        method_match && url_match
+      })
+      .expect("Mock not found");
+    // Assert upstream request
+    let upstream_request = mock.0.clone();
+    assert!(
+      self.expected_upstream_requests.contains(&upstream_request),
+      "Unexpected upstream request: {:?}",
+      upstream_request
+    );
+
+    // Clone the response from the mock to avoid borrowing issues.
+    let mock_response = mock.1.clone();
+
+    // Build the response with the status code from the mock.
+    let status_code = reqwest::StatusCode::from_u16(mock_response.0.status)?;
+    let mut response = Response { status: status_code, ..Default::default() };
+
+    // Insert headers from the mock into the response.
+    for (key, value) in mock_response.0.headers {
+      let header_name = HeaderName::from_str(&key)?;
+      let header_value = HeaderValue::from_str(&value)?;
+      response.headers.insert(header_name, header_value);
+    }
+
+    // Set the body of the response.
+    response.body = ConstValue::try_from(serde_json::from_value::<Value>(mock_response.0.body)?)?;
+
+    Ok(response)
+  }
+}
+
+#[tokio::test]
+async fn test_body_yaml() {
+  let spec = HttpSpec::read("tests/data/sample.yaml").unwrap();
+  assert_downstream(spec).await;
+}
+
+async fn assert_downstream(spec: HttpSpec) {
+  let has_only_annotation = spec
+    .downstream_assertions
+    .iter()
+    .any(|assertion| matches!(assertion.annotation, Some(Annotation::Only)));
+
+  for downstream_assertion in spec.downstream_assertions.iter() {
+    match &downstream_assertion.annotation {
+      Some(Annotation::Skip) if !has_only_annotation => {
+        let request_details = format_request_details(&downstream_assertion.request.0);
+        println!("Skipping test in: {}\nRequest Details: {}", spec.name, request_details);
+        continue;
+      }
+      Some(Annotation::Only) | None
+        if !has_only_annotation || matches!(downstream_assertion.annotation, Some(Annotation::Only)) =>
```

Comment:
The filtering of `only` `skip` `fail` should happen across all files, not just the downstream assertion.

Reply:
Done